module-apps/apps-common/ApplicationCommon.cpp

```diff
@@ -772,243 +801,113 @@ namespace app
         sender->bus.sendUnicast(msg, application);
     }
 
-    void ApplicationCommon::handlePhoneModeChanged(sys::phone_modes::PhoneMode mode)
+    void ApplicationCommon::actionPopupPush(std::unique_ptr<gui::SwitchData> params)
     {
-        auto flightModeSetting = settings->getValue(settings::Cellular::offlineMode, settings::SettingsScope::Global);
-        bool flightMode        = flightModeSetting == "1" ? true : false;
-
-        using namespace gui::popup;
-        const auto &popupName = resolveWindowName(gui::popup::ID::PhoneModes);
-        if (const auto currentWindowName = getCurrentWindow()->getName(); currentWindowName == popupName) {
-            updateCurrentWindow(std::make_unique<gui::ModesPopupData>(mode, flightMode));
+        auto rdata = dynamic_cast<gui::PopupRequestParams *>(params.get());
+        if (rdata == nullptr) {
+            assert(0 && "this should never happen");
+            return;
         }
-        else {
-            switchWindow(popupName, std::make_unique<gui::ModesPopupData>(mode, flightMode));
+        // two lines below is to **not** loose data on copy on dynamic_cast, but to move the data
+        (void)params.release();
```

Comment:
??

Reply:
this is a rare moment when I've added comments at above - is it insufficient? 
there is no `dynamic_pointer_cast` which would enable us to dynamic_cast unique_ptr with no information loss. As it would trim tada to gui::PopupRequestParams. 

Issue here is that we operate everywhere - on SwitchData even though we require gui::PopupRequestParams everywhere. But it was faster/more convenient for someone to implement it the way it is. Therefore spaghetti :)

Reply:
isn't `statc_cast` applicable here?